### ISSUE TITLE
Support for libgpiod as pin factory?

### ISSUE THREAD

**Issue by fdanapfel** [gpiozero/gpiozero#840]
Are there any plans to add support for [libgpiod](https://github.com/brgl/libgpiod) as a pin factory, which is the current recommended way of accessing the GPIO pins on the Raspberry Pi (see for example https://blog.adafruit.com/2018/11/26/sysfs-is-dead-long-live-libgpiod-libgpiod-for-linux-circuitpython/)?

The current supported pin factories only seem to be working on Raspian based systems, because only Raspian supports the 'old' style mechanism of using sysfs or /dev/gpiomem to access GPIO pins. Whereas other Linux Distributions like Fedora only support the 'new' style of using libgpiod, which currently makes gipozero unuseable on non-Raspian environments.

**Comment by bennuttall** [comment]
No plans but it sounds like an obvious next step. How feature complete is libgpiod?

The way gpiozero works is through "[pin factories](https://gpiozero.readthedocs.io/en/latest/api_pins.html)" which rely on underlying pin libraries. We currently support RPi.GPIO, pigpio (also RPIO which isn't maintained any more and only works on Pi 1), and an experimental native driver. If it's possible to write a pin factory for libgpiod ([see the others for example](https://github.com/gpiozero/gpiozero/tree/master/gpiozero/pins)) then there's no reason we couldn't add it.

**Comment by lurch** [comment]
See also https://github.com/gpiozero/gpiozero/issues/702#issuecomment-503370895
(but I've no idea if or what might have changed in the meantime)

**Comment by bennuttall** [comment]
Yes that did cross my mind but the issue has gone.

**Comment by fdanapfel** [comment]
> No plans but it sounds like an obvious next step. How feature complete is libgpiod?

I'm not an expert on this, but since libgpiod is part of the mainline kernel and seems to be [actively maintained](https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git/) I would hope that it is fairly complete.
 
> The way gpiozero works is through "[pin factories](https://gpiozero.readthedocs.io/en/latest/api_pins.html)" which rely on underlying pin libraries. We currently support RPi.GPIO, pigpio (also RPIO which isn't maintained any more and only works on Pi 1), and an experimental native driver. If it's possible to write a pin factory for libgpiod ([see the others for example](https://github.com/gpiozero/gpiozero/tree/master/gpiozero/pins)) then there's no reason we couldn't add it.

There are [python bindings](https://github.com/brgl/libgpiod/tree/master/bindings/python) available for libgpiod so I would assume writing a pin factory for libgpiod shouldn't be too difficult (althought tt would be way over the level of my python skills).

**Comment by lurch** [comment]
It's a shame there's no documentation for the python bindings :confused:

**Comment by bennuttall** [comment]
> There are python bindings available for libgpiod so I would assume writing a pin factory for libgpiod shouldn't be too difficult (althought tt would be way over the level of my python skills).

If you can flash an LED on a Raspberry Pi using that Python library, in the distro of your choice, and document the process (e.g. use this Fedora image, install this library from here, run this python file) so we have something to work with, that would be immensely helpful.

**Comment by bennuttall** [comment]
by "document this process" I just mean a few bullet points in this issue.

**Comment by bennuttall** [comment]
This project might be useful for reference: https://github.com/aoertel/rpi-rf-gpiod

**Comment by bennuttall** [comment]
python3-libgpiod exists in Debian Buster: https://packages.debian.org/buster/python3-libgpiod

but is it in Raspbian?

**Comment by ukBaz** [comment]
A presentation my search engine found for me on this topic:
https://ostconf.com/system/attachments/files/000/001/532/original/Linux_Piter_2018_-_New_GPIO_interface_for_linux_userspace.pdf?1541021776

Contains:
![image](https://user-images.githubusercontent.com/8608260/75691169-d9ae8180-5c9b-11ea-9b09-aa289e6f8e1a.png)

**Comment by fdanapfel** [comment]
> If you can flash an LED on a Raspberry Pi using that Python library, in the distro of your choice, and document the process (e.g. use this Fedora image, install this library from here, run this python file) so we have something to work with, that would be immensely helpful.

OK, here's what I've tried so far:
- downloaded an image of the current Fedora 31 for armhfp from https://arm.fedoraproject.org and put it on a SD card
- booted a Raspberry Pi 3 B+ using this Fedora 31 SD card and ran through the initial configuration (wifi, user creation etc.)
- installed libgpiod userspace utils  and the python3 bindings for it:
   `sudo dnf install libgpiod-utils python3-libgpiod`
- ran the gpioset.py script from https://github.com/brgl/libgpiod/tree/master/bindings/python/examples to verify that activating  a LED connected to one of the GPIO pins works

I was also able to install gpiozero by running
`sudo dnf install python3-gpiozero`
but as mentioned in my initial comment it doesn't work since none of the supported pin factories work on Fedora 31.

I've also discovered that somebody is working on a reimplementation of RPI.GPIO using libgpiod: https://github.com/underground-software/python3-libgpiod-rpi
But it seems to be missing some functionality needed by gpiozero, since even trying to use a basic gpiozero python script to activate a LED results in a python traceback.

**Comment by lurch** [comment]
> I've also discovered that somebody is working on a reimplementation of RPI.GPIO using libgpiod: https://github.com/underground-software/python3-libgpiod-rpi

Oooh, that's cool! :heart:  ping @theyoyojo in case they have any comments they'd like to add here :slightly_smiling_face:

**Comment by theyoyojo** [comment]
Hi everyone,

Thanks for the ping @lurch.

@fdanapfel Can you provide more detail on your error? I would be happy to work with you and anyone else who is interested to get gpiozero to to work with our library.

As a bit of background, I am actively developing this library with group of students at the University of Massachusetts Lowell of which I am a member. 

Our goal with python3-libgpiod-rpi is to develop a distro-generic lightweight replacement for RPi.GPIO that uses the standard kernel interface instead of relying on Raspbian-specific patches (i.e. `mmap("/dev/gpiomem")`) or insecure kernel configuration (i.e. `mmap("/dev/mem")`). In doing so, we hope to preserve the utility of the many existing tutorials and libraries that rely on the RPi.GPIO API, such as gpiozero, while eliminating the non-modular and somewhat disorganized codebase of the original library.

We welcome any feedback, suggestions, or criticism.

**Comment by fdanapfel** [comment]
> @fdanapfel Can you provide more detail on your error? I would be happy to work with you and anyone else who is interested to get gpiozero to to work with our library.

@theyoyojo, thanks for your support with this!  

Here's the error I'm getting when trying to run a very simple test wit gpiozero to turn on a led using the python-libgpiod-rpi as the pin factory for gpiozero:
`# ./gpiozero-test.py 
/usr/lib/python3.7/site-packages/gpiozero/devices.py:452: PinFactoryFallback: Falling back from rpigpio: module 'RPi.GPIO' has no attribute 'I2C'
  'Falling back from %s: %s' % (name, str(e))))
/usr/lib/python3.7/site-packages/gpiozero/devices.py:452: PinFactoryFallback: Falling back from rpio: No module named 'RPIO'
  'Falling back from %s: %s' % (name, str(e))))
/usr/lib/python3.7/site-packages/gpiozero/devices.py:452: PinFactoryFallback: Falling back from pigpio: No module named 'pigpio'
  'Falling back from %s: %s' % (name, str(e))))
/usr/lib/python3.7/site-packages/gpiozero/devices.py:452: PinFactoryFallback: Falling back from native: unable to open /dev/gpiomem or /dev/mem; upgrade your kernel or run as root
  'Falling back from %s: %s' % (name, str(e))))
Traceback (most recent call last):
  File "./gpiozero-test.py", line 2, in <module>
    from gpiozero import LED
  File "/usr/lib/python3.7/site-packages/gpiozero/__init__.py", line 22, in <module>
    from .devices import (
  File "/usr/lib/python3.7/site-packages/gpiozero/devices.py", line 486, in <module>
    Device.pin_factory = _default_pin_factory()
  File "/usr/lib/python3.7/site-packages/gpiozero/devices.py", line 453, in _default_pin_factory
    raise BadPinFactory('Unable to load any default pin factory!')
gpiozero.exc.BadPinFactory: Unable to load any default pin factory!
`

This is the test script:
`# cat gpiozero-test.py 
#!/usr/bin/env python3
from gpiozero import LED
import os 

ledPin = 14
led = LED(ledPin)

led.on()
`

And there are the GPIO2 related packages I have installed (python3-libgpiod-rpi-0.2-1.armv7hl was built using the spec file from your GitHub repo):
`# rpm -qa|grep gpio
libgpiod-utils-1.5-1.fc31.armv7hl
python3-gpiozero-1.4.1-3.fc31.noarch
python3-libgpiod-rpi-0.2-1.armv7hl
libgpiod-1.5-1.fc31.armv7hl
python3-libgpiod-1.5-1.fc31.armv7hl`

**Comment by lurch** [comment]
> PinFactoryFallback: Falling back from rpigpio: module 'RPi.GPIO' has no attribute 'I2C' 

Looks like that's because of https://github.com/gpiozero/gpiozero/blob/master/gpiozero/pins/rpigpio.py#L101 but I2C isn't (yet) used anywhere in gpiozero.

**Comment by bennuttall** [comment]
I guess it's more that the `RPi.GPIO` clone doesn't support I2C yet, but gpiozero expects to be able to import the `I2C` constant. If you set the constant (it's 42), maybe it'll work (or at least get further).

**Comment by AndyWas** [comment]
This sounds like excellent news, and if I can do anything to help with testing etc. I am happy to oblige.  I am trying to rework a Raspbian / RPi.GPIO project on Fedora Server 31 / RPi 3B+ and have hit exactly this issue.  I tried building pigpiod from source - apparently successfully - but it segfaults as soon as I run it, and the maintainer has made it clear no help will be forthcoming ([https://github.com/joan2937/pigpio/issues/340](url)).

I'm not a hardcore programmer by any means but I'm good at working through things to find out where they are going wrong.

**Comment by bennuttall** [comment]
I presume gpiozero itself "works" on Fedora (there's no reason it wouldn't - it's pure Python). 

Try installing gpiozero with pip, setting `GPIOZERO_PIN_FACTORY=mock` and trying some GPIO code, which should "work". (See [mock pins](https://gpiozero.readthedocs.io/en/latest/api_pins.html#mock-pins)). Alternatively, try running the test suite. See [tests](https://gpiozero.readthedocs.io/en/latest/development.html#test-suite).

And with the pigpio client installed, it should also work remotely controlling a Pi running Raspian. But again, that doesn't really prove anything. Just that Python works.

However, getting a supported GPIO back-end to work is presumably the hard part. I guess RPi.GPIO is largely focused on Debian/Raspbian - will it not work at all in Fedora? And the next thing to try is pigpio. So yeah, I see where you've got to. And there's not really anywhere else to go (RPIO is dead, native pin won't work) before you get to libgpiod.

@funnelfiasco anyone in Fedora working on (or would be willing to investigate) getting Raspberry Pi GPIO libraries working?

**Comment by funnelfiasco** [comment]
> anyone in Fedora working on (or would be willing to investigate) getting Raspberry Pi GPIO libraries working?

I'll find out!

**Comment by AndyWas** [comment]
> I guess RPi.GPIO is largely focused on Debian/Raspbian - will it not work at all in Fedora?

I think RPi.GPIO / RPIO are non-starters on Fedora due to the way that /proc/cpuinfo is used to determine the hardware and revision of the Raspberry Pi: [https://bugzilla.redhat.com/show_bug.cgi?id=1471731](url).  I guess this is one of the "Raspbian-specific patches".  Neither of those projects seem to be very active any more either.

**Comment by lurch** [comment]
> I think RPi.GPIO / RPIO are non-starters on Fedora due to the way that /proc/cpuinfo is used to determine the hardware and revision of the Raspberry Pi

I believe the "modern" way to do that is via DeviceTree e.g. https://github.com/gpiozero/gpiozero/blob/master/gpiozero/pins/local.py#L93

**Comment by AndyWas** [comment]
> I believe the "modern" way to do that is via DeviceTree

That may very well be so; however I can't see anyone now modifying RPi.GPIO or RPIO that way for the benefit of distros other than Raspbian.  And in any case it would still be using the deprecated sysfs or /dev/gpiomem methods.  Getting libgpiod to work as a backend seems a much better idea, especially hiding it in the flayed skin of RPi.GPIO.

**Comment by funnelfiasco** [comment]
Here's the response about [support in Fedora](https://lists.fedoraproject.org/archives/list/arm@lists.fedoraproject.org/message/TDRPXYIEFWLES64CBWJYDPTWMTYUBLI3/):

> They need to be ported to use libgpiod and associated
bindings so they use the modern and secure interface, not the old
deprecated interfaces.
>
> There was discussion about forking them or implementing the same
interface in a different library but I don't see the point in that
because it means all the millions of instructions on the internet
won't work. What needs to be doing is to engage with the maintainer of
RPi.GPIO to get the support landed there so it uses libgpiod by
default and if that's not available falls back to the old way, that
way it will work with new systems buy fall back and all the
instructions for cool projects will just work everywhere.

**Comment by bcroston** [comment]
I am the maintainer of RPi.GPIO.  I've not done any work on libgpiod yet.  There will be quite a performance penalty compared with /dev/gpiomem.

**Comment by AndyWas** [comment]
OK so I have used pip to uninstall RPi.GPIO and RPIO, and downloaded and installed @theyoyojo's latest "Exodus" RPM of python3-libgpiod-rpi.  **Can anyone please advise how to install that in Python so I can try using it from GPIO Zero as a replacement for the RPi.GPIO pin factory?**  The project website explicitly says not to pip install RPi.GPIO as that will simply reinstall the original module.

**Comment by bennuttall** [comment]
I'm afraid it's not as easy as that. gpiozero provides the means to use different pin libraries underneath - but the implementation of a pin factory is required. See how others are implemented here: https://github.com/gpiozero/gpiozero/tree/master/gpiozero/pins

See the discussion earlier in this thread

**Comment by AndyWas** [comment]
Hi @fdanapfel 
Can you tell me what you did to install python-libgpiod-rpi and configure it as the pin factory for gpiozero please?

**Comment by fdanapfel** [comment]
@AndyWas, all I did is described in my previous posts. 

But as you can see there I did not get it to work because there seem to be some parts missing in python-libgpiod-rpi which prevent it from being used as a pin factory for gpiozero.

**Comment by I-Programmer-info** [comment]
I think there may be another potential problem.
The edge event handling in GPIO Zero uses Sysfs which is deprecated and scheduled for removal this year to be replaced by libgpiod. This means that irrespective of any new pin factory based on libgpiod the existing pin factories need to be updated to use it.

**Comment by bennuttall** [comment]
I'd imagine the pin factories wouldn't need updating, but the libraries themselves will need updating. I guess the native pin factory would need updating (but that's not a library, it's just implemented in the pin factory) or even deprecating.

**Comment by I-Programmer-info** [comment]
The event mixin?
Does the use of sysfs go deeper?
And don't deprecate the native pin factory if at all possible - its an excellent teaching aid.

**Comment by bennuttall** [comment]
The `EventsMixin` just wraps the callbacks and event changes, relative to the way the pin factory does events, nothing to do with sysfs: https://github.com/gpiozero/gpiozero/blob/master/gpiozero/mixins.py#L182

**Comment by I-Programmer-info** [comment]
My mistake - its rpigpio that uses sysfs to implement events..

**Comment by lurch** [comment]
There's so much laying within pin libraries / pin factories / etc. that it's an easy mistake to make :wink:

**Comment by I-Programmer-info** [comment]
So true... its one of the advantages of the native pin factory i.e. that all the code is there and part of the same documentation.
And I guess we now have to worry about rpigpio updating its code to not use sysfs.

**Comment by lurch** [comment]
> And I guess we now have to worry about rpigpio updating its code to not use sysfs.

ping @bcroston to see if this is something already on his radar?

**Comment by I-Programmer-info** [comment]
So true... its one of the advantages of the native pin factory i.e. that all the code is there and part of the same documentation.
And I guess we now have to worry about rpigpio updating its code to not use sysfs.

**Comment by theyoyojo** [comment]
Hi @AndyWas,

I just tested your script with RPi.GPIO2 0.3.0a2 (the PyPi name for python3-libgpiod-rpi) and it works with gpiozero when patched with the following:

```
diff --git a/gpiozero/pins/local.py b/gpiozero/pins/local.py
index 1c40c47..3bca73d 100644
--- a/gpiozero/pins/local.py
+++ b/gpiozero/pins/local.py
@@ -61,6 +61,10 @@ from ..devices import Device, SharedMixin
 from ..output_devices import OutputDevice
 from ..exc import DeviceClosed, PinUnknownPi, SPIInvalidClockMode
 
+REVISION_TABLE = {
+    "Raspberry Pi 3 Model B": hex(0xa22082),
+}
+
 
 class LocalPiFactory(PiFactory):
     """
@@ -98,6 +102,13 @@ class LocalPiFactory(PiFactory):
         except IOError as e:
             if e.errno != errno.ENOENT:
                 raise e
+        try:
+            with io.open('/proc/device-tree/model', 'r') as f:
+                modelstr = f.readline()[:-1]
+                revision = REVISION_TABLE[modelstr]
+        except (IOError, KeyError) as e:
+            if not isinstance(e, IOError) or e.errno != errno.ENOENT:
+                raise e
             with io.open('/proc/cpuinfo', 'r') as f:
                 for line in f:
                     if line.startswith('Revision'):
```

The blocking issue preventing out-of-the-box compatibility between gpiozero and python3-libgpiod-rpi is the method of hardware revision detection used gpiozero.

**Comment by joan2937** [comment]
I have written a couple of Python modules which provide functionality via the /dev/gpiochip interface.

They have had little testing but appear reasonably stable.

[lgpio](http://abyz.me.uk/lg/py_lgpio.html) allows control of a local Pi's GPIO, i.e. it is like RPi.GPIO and RPIO.GPIO.  It is auto-generated from a C library using SWIG.

[rgpio](http://abyz.me.uk/lg/py_rgpio.html) allows control of local and remote Pi's GPIO, i.e. it is like pigpio.  Like pigpio it also uses a client server module.

I say Pis but these are both general Linux libraries and do not use any special Pi feature, i.e. they should tun on any Linux machine.

lgpio and rgpio share the following features

- reading and writing GPIO singly and in groups
- software timed PWM and waves
- GPIO callbacks
- pipe notification of GPIO alerts
- I2C wrapper
- SPI wrapper
- serial link wrapper

In addition the rgpio Python module

- can run on Windows, Macs, or Linux
- controls one or more SBCs
- simple file handling
- creating and running scripts on the rgpiod daemon
- has an optional access control system giving fine-grained control over which GPIO, I2C bus, SPPI bus etc. can be controlled by individual users.

https://github.com/joan2937/lg

**Comment by tedsluis** [comment]
> I have written a couple of Python modules which provide functionality via the /dev/gpiochip interface.
> 
> They have had little testing but appear reasonably stable.
> 
> [lgpio](http://abyz.me.uk/lg/py_lgpio.html) allows control of a local Pi's GPIO, i.e. it is like RPi.GPIO and RPIO.GPIO. It is auto-generated from a C library using SWIG.
> lgpio and rgpio share the following features
> 
> * reading and writing GPIO singly and in groups
> * software timed PWM and waves
> * GPIO callbacks
> * pipe notification of GPIO alerts
> * I2C wrapper
> * SPI wrapper
> * serial link wrapper
> https://github.com/joan2937/lg

Very useful since Fedora on raspberry does not support the legacy/sys/class/gpio interface anymore! I have used your lgpio python module to PWM control the fans of my raspberry pis on fedora 33, see https://github.com/tedsluis/raspberry-pi-pwm-fan-control  Great python modules! Thnx!

**Comment by waveform80** [comment]
Just a quick follow-up on this; I've been working on packaging lgpio/rgpio for Ubuntu and RaspiOS - that's nearly complete (although there's still some pending bits for rgpiod to do) - current testing packages available from https://launchpad.net/~waveform/+archive/ubuntu/lg and I've just finished a PR adding lgpio as a built-in pin library which is #927. I've got another branch where I'm working on rgpio support (as another built-in pin library).

In case anyone's wondering why I targetted lgpio in preference to libgpiod it's to do with facilities: lgpio provides full support for ... well, pretty much everything including (software-based) PWM. libgpiod provides support for the bare-bones of GPIO facilities only (so no PWM ... well, unless you want to try flipping pins high and low as fast as you can from Python ...). So lgpio will serve as a pretty full featured alternative to RPi.GPIO, but libgpiod cannot.

However, it would be a nice fallback (like NativePin currently is) and I think it's vaguely important to have a pure Python implementation (like NativePin) covering libgpiod (if only because it's proved useful for various people who don't know C to familiarize themselves with talking to GPIO pins straight from Python). So, once the rgpio pin implementation is done, I intend to spend a couple of days writing a NativePin-esque pin implementation based purely on libgpiod. It *won't* be fully-featured (in other words, no PWM, just like NativePin) but I figure that's okay given lgpio (which also uses the gpiochip devices) will be there too (above it in the preference order), and it'll still be a useful fallback.

**Comment by matthijskooijman** [comment]
> However, it would be a nice fallback (like NativePin currently is) and I think it's vaguely important to have a pure Python implementation (like NativePin) covering libgpiod 

An additional reason is that, AFAICS, installing lgpio is currently a manual affair.

The [lgpio pypi page](https://pypi.org/project/lgpio/) says it cannot be installed througy pypi, AFAIU because pypi cannot compile the native parts properly. I'm not sure that pip really cannot install native/compiled stuff, but it doesn't work currently in any case (if you just `pip install lgpio`, it seems to just install an empty and non-functional package, which leads to `PinFactoryFallback: Falling back from lgpio: module 'lgpio' has no attribute 'BOTH_EDGES'`). It seems Ubuntu [has a precompiled package for lgpio starting with 21.04](https://packages.ubuntu.com/search?keywords=lgpio&searchon=names&suite=all&section=all), but [not Debian](https://packages.debian.org/search?keywords=lgpio&searchon=names&suite=all&section=all) for some reason.

I have recently done a few things with libgpiod from python and noticed there are currently two very similar python bindings:
 1. The bindings that are shipped with the libgpiod C-library itself. This is a wrapper around the actual C library and is what you get when you install the `python3-libgpiod` package. Best documentation I've found is the [examples](https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git/tree/bindings/python/examples) combined with the C libgpiod API documentation (i.e. `libgpiod-doc` on Debian).
 2. [Python3-gpiod](https://pypi.org/project/gpiod/), which is a pure python implementation of libgpiod, that talks directly to the device file. Its API mimics the libgpiod API.
 
As a heads up: because APIs are similar, and both use `import gpiod`, it's easy to confuse these two libraries (I was at some point using docs for one to write code for the other, which almost but not quite worked :-p).

**Comment by waveform80** [comment]
@matthijskooijman I'm in the process of attempting to get lg into Debian (based on my work on getting it into Ubuntu), but it's a slow process.

The full reasons for preferring lgpio are in the blog post Ben linked to but it basically boils down to: libgpiod has no PWM support (so no PWMLED, Servo, Motor support, etc.). Oh, and its documentation is ... well, I'll be generous and say "minimal". I do intend to write a libgpiod backend at some point. When I get around to it it'll be targetting the python3-libgpiod package for the simple reason that whatever we do eventually has to wind up packaged in raspios, ubuntu, and debian so going with dependencies that already exist there is always preferable.

**Comment by matthijskooijman** [comment]
Awesome blogpost, great to see an overview and history like that. Thanks for linking to, and writing it. It seems that most if not everything I posted above you already knew :-)

> Oh, and its documentation is ... well, I'll be generous and say "minimal".

Yeah, that's also my experience sadly.

> I do intend to write a libgpiod backend at some point. When I get around to it it'll be targetting the python3-libgpiod package for the simple reason that whatever we do eventually has to wind up packaged in raspios, ubuntu, and debian so going with dependencies that already exist there is always preferable.

Sounds good!

Thanks for following up!

**Comment by Gadgetoid** [comment]
I'm looking into lgpio and it's fine as a choice, but not an ideal package to be relying on as the cornerstone for GPIO on Raspberry Pi (and perhaps even the wider SBC ecosystem). Chiefly it's abdicating the responsibility of Python bindings to SWIG, lacks a Python PyPi distribution (making it impossible to use in virtual environments without exposing system site packages) and seems to include Serial, I2C and SPI wrappers, which are already solved problems in Python. That's a _lot_ of baggage for something that - as far as I can tell - just needs to be "libgpiod but with PWM."

Am I missing something? Is anything other than soft-PWM blocking a libgpiod-based backend from being full-featured? Is there any reason why a backend couldn't depend upon libgpiod *and* a dedicated libgpiod-based soft-PWM library, so dependencies are succinct and comprehensible to mere mortals?

**Comment by waveform80** [comment]
I agree the packaging situation with `lgpio` isn't ideal but I wasn't aware it was any better with `libgpiod`. However, I've just checked and someone has (this year!) uploaded a [PyPI package for `libgpiod`](https://pypi.org/project/libgpiod/#history) so I guess it would be easier at this point.

One additional reason I had for favouring `lgpio` is related to "remote" GPIO. The `pigpio` backend is likely to become obsolete (in a somewhat similar manner to `RPi.GPIO` given it also bangs on the hardware, albeit in rather more advanced ways), and it's currently the only method of achieving "remote" GPIO (which we've got a whole chapter on in the docs). Rather than throw that away, I'd like to add an `rgpio` implementation (which is verrrry similar to the `lgpio` implementation, unsurprisingly given it's from the same source), but (as ever) I've not had time yet (though I think I've got a partial branch for it lying around somewhere).

The I2C/SPI stuff becomes "interesting" in the remote case (and is another area where rgpio basically "solves everything" by just providing interfaces for that).

One final note I should add: PWM isn't mandatory to a pin implementation -- our "native" implementation doesn't support it and just throws an error when anyone tries to use PWM, so there's absolutely nothing (other than lack of time) preventing a `libgpiod` implementation right now; we could always tag on PWM support at some later date, but without PWM, we'd never set it at the top of the "default priority list" of implementations to attempt.

**Comment by waveform80** [comment]
Oh ... no, there's [another one](https://pypi.org/project/gpiod/) which is earlier. Is that the official one? Seems to match the name given in the [documentation](https://wiki.loliot.net/docs/lang/python/libraries/gpiod/python-gpiod-about/). I guess I'll figure that out when I get around to writing the layer!

**Comment by Gadgetoid** [comment]
@waveform80 ah, thank you for clarifying. I've never actually had cause to use the remote stuff, so it's not been on my radar. It suspect- then- that gpiozero is fundamentally not the right answer for most of the products I work with. Though it has a very particular affinity for products like Explorer HAT- since the two libraries solve the "make GPIO sort-of user-friendly" problem in their own respective ways.

The lack of (an up-to-date) `lgpio` in pypi is a nuisance when using a virtualenv- the defacto standard now on Bookworm OS- since you need to remember to specify `--system-site-packages` or suffer cryptic error messages. I think that particular pitfall is what prompted this PR https://github.com/gpiozero/gpiozero/pull/1063

As for the stuff on pypi-

The `gpiod` python package is implemented in pure-python, simply interfacing with the character device. It seems to be maintained, at least (source here: https://github.com/hhk7734/python3-gpiod), but not in any meaningful way official.

It looks like the `libgpiod` is an unofficial, drive-by upload of ... something... its contents - a bunch of pure Python resembling `gpio` - do not match those of `python3-libgpiod` which is a C library.

**Comment by waveform80** [comment]
Ah, something like #1063 should be landing fairly soon in GPIO zero (I'm afraid I couldn't include that one directly as there's a whole mess of other changes that'll be landing with it, including shifting the default pin library to lgpio for bookworm, and it was easier than sorting out all the git conflicts).

@bennuttall is there anything we can do about the lgpio packaging situation on PyPI? The point about venv's being mandatory in bookworm is rather important, and likely to impact anybody using them (which is probably quite a few people!)

**Comment by bennuttall** [comment]
I put pigpio on PyPI but it looks like joan has done lgpio herself. We can prompt an update or I can offer to take it on if she wants.

**Comment by Gadgetoid** [comment]
It might be that package needs some finagling to make PiWheels build the C extension.

Now you've got me wondering how much stuff I can make work with remote 😆

**Comment by joan2937** [comment]
@bennuttall 

Ben it would be very useful to me if you could look after the packaging of lgpio and/or rgpio.

When I originally did lgpio I could find no way of adding a Linux library.  Hopefully times have changed.

I have sent a pypi collaborator request for use as maintainer.  If you are happy with this I would like to change ownership of the packages to you and remove myself (I like Python as a tool but I am thinking of C when I use it).

**Comment by Gadgetoid** [comment]
> I like Python as a tool but I am thinking of C when I use it

I figured something like this was the case. I have the dubious honour of being mediocre at both, but biasing toward Python.

> When I originally did lgpio I could find no way of adding a Linux library. Hopefully times have changed.

Python packaging has probably changed completely ten times by now :laughing: But it's always been doable, it's just always been... [quite horrible to actually accomplish](https://github.com/WiringPi/WiringPi-Python/blob/master/setup.py). *cough cough static everything cough cough* This factoid was one of the pillars of WiringPi's inevitable demise since the static-linked abomination shipped in the Python package would inevitably fall behind the official C library and cause excess support headaches. I don't think the ecosystem really offers a good fix for this- at least not now venvs are the norm.

@bennuttall you've probably already run circles around me with PiWheels debugging, but if you need any help I've run the SWIG-bindings-as-a-Python-library gauntlet before with WiringPi and rpi_ws281x.

**Comment by bennuttall** [comment]
Thanks both!

**Comment by joan2937** [comment]
@Gadgetoid

I remember SWIG had a bit of a learning curve.  I dread to think of the learning curve needed to use Python bindings directly.  I'd probably have given up in frustration.

**Comment by bennuttall** [comment]
@Gadgetoid I've managed to build it by modifying the setup.py (see https://github.com/bennuttall/lg/blob/pypi/PY_LGPIO/setup.py#L12) but it required me to *install* lg on my machine first. I'll give you push rights to my fork if you want to have a play.

**Comment by Gadgetoid** [comment]
> it required me to _install_ lg on my machine first

Based on my experience with WiringPi, I think this is an acceptable compromise, though I think there are patterns that give the user the choice to pull in a packaged copy of the library (via pypi, ugh) versus using the system library.

What's the state of binary distributions in Python these days? PiWheels has sort of removed my need to think about it- can it handle building the bindings and an (optionally installed) copy of the "lg" library? Would those actually get correctly distributed. Sheesh if `lg` depends on `lgpiod` then does that just kick the can down the road anyway?

The next hurdle, perhaps, is that `setup.py` is basically obsolete so ideally this should be migrated to whatever the standard approach for Python packaging is today lest we suffer a wall of deprecation warnings. :grimacing: 

Sorry I've been a little in the dark with Pi stuff for the last few years, since Pico ate all my time. I usually take pains to own as much of the hardware interfacing process as I can :laughing: - I'm coming for GPIO once I figure out how it should work. (I accidentally adopted the sysfs gpio library before I realized it was the wrong one :laughing:)

**Comment by Gadgetoid** [comment]
@bennuttall these changes make PY_LGPIO build a completely standalone library that can be installed into a venv without any system lg install - https://github.com/joan2937/lg/compare/master...gpiozero:lg:pypi-standalone

**Comment by Gadgetoid** [comment]
> I agree the packaging situation with `lgpio` isn't ideal but I wasn't aware it was any better with `libgpiod`. However, I've just checked and someone has (this year!) uploaded a [PyPI package for `libgpiod`](https://pypi.org/project/libgpiod/#history) so I guess it would be easier at this point.

@waveform80  Okay I reached out to this "someone" and made a total fool out of myself by not realizing that they are- in fact- *the maintainer of libgpiod*

They put the package up to serve their own purposes - building with Yocto - and have no immediate interest in making it work cross-platform, nor should we necessarily expect that from them. It's not easy or necessarily even the right thing to do.

They mentioned it might be possible, in the short term, to have it fail a *little more gracefully* when you try to install it. Right now you'll notice a lack of PiWheels for it, and - upon trying to install - it'll baulk with a wall of errors relating to a mismatch between the library and the system libgpio headers/library.

So I tentatively offered to explore doing what I did with lgpio / "PY_LGPIO" and creating a new libgpiod package that packages up the library dependencies alongside the actual C Python library, and asked if they would consider releasing it as "libgpiod".

Soo- the immediate state of libgpiod/gpiod is a bit of a brick-wall, but there's a ray of light for getting where we need.

I also ran some benchmarks between `libgpiod` and `gpiod` (both of which conflict for the `gpiod` module name in Python, ugh) and found `libgpiod` (the official library, installed via its Debian package) to be some 2-3x faster in a basic pin toggle test.

**Comment by Gadgetoid** [comment]
@bennuttall can you let me know if this builds on PiWheels? https://github.com/Gadgetoid/libgpiod-python/tree/2.0.0

Source dist here: https://github.com/Gadgetoid/libgpiod-python/releases/tag/2.0.0

**Comment by waveform80** [comment]
Sorry this has taken me so long to get back to -- mantic's release didn't go quite as smoothly as hoped!

I dug into this a bit in the evenings this week and here's where I *think* things stand...

On raspios bookworm:

* Installed python3-gpiozero from apt; worked out of the box on Pi 4 and 5
* Created venv *with* system site packages, forced a pip install of gpiozero within this, worked happily on Pi 4 and 5
* Created venv *without* system site packages, pip installed gpiozero within this, everything breaks on 5 because there's no default pin driver, and on 4 because native doesn't support PWM etc. etc.
* pip installing rpi.gpio within the venv works on 4, doesn't work on 5 (for the obvious reasons)
* pip installing lgpio within the venv on 5 doesn't work because the package is out of date

Now ... if we simply bump lgpio's version on PyPI that *should* solve things for bookworm, because it can still build successfully there because liblgpio1 is installed by default. I did have something ready for this a couple of weeks ago and then mantic's release got in the way ... I'll see if I can dig it out and get @bennuttall to push it.

But that doesn't fix things for bullseye, which has no packaging of liblgpio1. May be worth considering a backport of that library for those sticking on bullseye for a while (which is probably a fair number of people).

**Comment by Gadgetoid** [comment]
> Now ... if we simply bump lgpio's version on PyPI that _should_ solve things for bookworm, because it can still build successfully there because liblgpio1 is installed by default.

Is vendoring lgpio so we don't play system dependency bingo any time we want to build/install a Python package so bad?

I'm widely of the opinion that distro packages should be for *waves hand at vague distro shipped tooling here* and that userspace Python should comprise pretty much exclusively of system-independent pypi packages. I'd say the authors of "venv" agree with me, since `--system-site-packages` is *not* the default. Not doing this potentially hamstrings portability of user projects, which isn't great if we're trying to go the _"you can package up your project with requirements.txt and be reasonably sure someone else can get it working in a vent regardless of their system"_ route.

I guess it boils down to: how likely are there to be breaking API changes for things like remote GPIO? Since it's somewhat unique in that it relies upon a server?? at the other end for remote GPIO to function.

I am very far along the process of [vendoring libgpiod into its Python package](https://github.com/Gadgetoid/libgpiod/pull/5), and this technique is far from unprecedented.

**Comment by waveform80** [comment]
You may be right -- my experience of vendoring stuff is extremely mixed (looks sideways at snaps); sometimes it works, sometimes it just pushes the issue of breakage further down the stack (below the bit that's been vendored) and then the debugging gets really fun (and by fun I mean ... erm ... not!).

I'll have a look how easy it might be in lgpio's case and whether or not it's worth producing a couple of wheels (with and without vendoring of the lib).

**Comment by Gadgetoid** [comment]
Of course there are cases where a vendored library intended to interact with some system interface might be incompatible, but I'd hope we're well into the "don't break userspace" phase now... wait, is libgpiod considered kernel or userspace? :laughing: 

Anyway an official libgpiod that can actually be used in gpiozero is *coming* and if you somehow have the time (I know we're all absolutely up against the wall here, which is why I'm in full panic mode) to poke holes in my libgpiod patches I'd welcome your input. Despite my experience (or in some cases because of it) I'm keenly aware I am not The Ultimate Knowledge God.

As an aside "Phase 3" of this mad dash is to squeeze a PWM interface out of Pi and write a nice, concise, Python PWM library to abstract PIO/HW, HW and SW solutions into one, neat, tidy package. That's also something I wouldn't want to do in a vacuum.

**Comment by Gadgetoid** [comment]
For better or worse a brand shiny new gpiod for Python is available now- https://pypi.org/project/gpiod/

Bart has no interest in providing .whl files, since it’s an extra maintenance burden, but this should build on anything with a working compiler and python3-dev.

**Comment by warthog618** [comment]
Hi,

I found my way here as I was playing with **gpiozero/lgpio**, and the very first thing I did with it revealed an [issue](https://github.com/joan2937/lg/issues/26).  Anyone else experience that?

I've been waiting for a response from joan, but got bored so have written a **gpiod** pin factory so it is possible to do basic monitoring of a pin with **gpiozero** with the minimal CPU usage that I'm used to.  If there is any interest in that I can tidy it up and create a PR.

Of course it only does **DigitalInputDevice** and **DigitalOutputDevice**, no **PWMOutputDevice**, but **libgpiod/gpiod** has no business doing PWM, so you get that.  OTOH this factory does debounce in the kernel, whereas **lgpio** does it in userspace, so you win some, you lose some.

Oh, and can someone fix the **lgpio** factory - it ignores the chip param and only supports the automagically selected chip.  Seems @waveform80 broke it in e3c73ee5.  I could raise an issue and a PR for that, but it is a one-liner, so can someone just fix it?  As it stands it is impossible to point a **lgpio** factory at a **gpio-sim** created gpiochip.  How do you do your testing?

**Comment by waveform80** [comment]
> Hi,
> 
> I found my way here as I was playing with **gpiozero/lgpio**, and the very first thing I did with it revealed an [issue](https://github.com/joan2937/lg/issues/26). Anyone else experience that?
> 
> I've been waiting for a response from joan, but got bored so have written a **gpiod** pin factory so it is possible to do basic monitoring of a pin with **gpiozero** with the minimal CPU usage that I'm used to. If there is any interest in that I can tidy it up and create a PR.

Sure, always happy to take new pin factories (in fact, they don't technically need to be part of gpiozero given we use entry-points -- any package can declare new pin factories which will be discovered via the [gpiozero_pin_factories](https://github.com/gpiozero/gpiozero/blob/6bfab8aa52c4098503589c3cd0bff7371a598ff4/setup.cfg#L52C1-L52C23) entry-point.

> Of course it only does **DigitalInputDevice** and **DigitalOutputDevice**, no **PWMOutputDevice**, but **libgpiod/gpiod** has no business doing PWM, so you get that. OTOH this factory does debounce in the kernel, whereas **lgpio** does it in userspace, so you win some, you lose some.

Yup, that's fine -- there's no hard requirement for supporting PWM in pin factories (the native one never has for instance).

> Oh, and can someone fix the **lgpio** factory - it ignores the chip param and only supports the automagically selected chip. Seems @waveform80 broke it in [e3c73ee](https://github.com/gpiozero/gpiozero/commit/e3c73ee5aa3f3035eb26a367c7fd53c5ef0e7309). I could raise an issue and a PR for that, but it is a one-liner, so can someone just fix it? As it stands it is impossible to point a **lgpio** factory at a **gpio-sim** created gpiochip. How do you do your testing?

Yeah, I need to fix that anyway given the moves afoot to migrate gpiochip4 to gpiochip0 in newer RaspiOS kernels (already patched rpi-lgpio). Testing is done with a [mock factory](https://github.com/gpiozero/gpiozero/blob/master/gpiozero/pins/mock.py) for the device classes, but with [physical tests](https://github.com/gpiozero/gpiozero/blob/6bfab8aa52c4098503589c3cd0bff7371a598ff4/tests/test_real_pins.py#L32) for the pin factories themselves (because there's no other reliable way of testing the libraries that just bash the GPIO registers). It's one of those things I've been promising myself to set up a runner for, but never find the time to manage.

**Comment by Gadgetoid** [comment]
> given the moves afoot to migrate gpiochip4 to gpiochip0 in newer RaspiOS kernels

Oh gosh, is there a citation for this? 😭 I think I'm okay because I rely on the labels.

**Comment by waveform80** [comment]
@Gadgetoid sure -- here you go: https://github.com/raspberrypi/linux/pull/6144

**Comment by waveform80** [comment]
@Gadgetoid oh, and here's what I did in rpi-lgpio in response: https://github.com/waveform80/rpi-lgpio/commit/bb7ff0e2072eb15ad81c713201a2616e69c464a8 which should "discover" the user-accessible GPIOs on both kernel versions (I need to get this into gpiozero too, but I'm horribly busy this week)

**Comment by warthog618** [comment]
> Sure, always happy to take new pin factories (in fact, they don't technically need to be part of gpiozero given we use entry-points -- any package can declare new pin factories which will be discovered via the [gpiozero_pin_factories](https://github.com/gpiozero/gpiozero/blob/6bfab8aa52c4098503589c3cd0bff7371a598ff4/setup.cfg#L52C1-L52C23) entry-point.
> 

Ok, I'm confused, so what is the purpose of this issue?
The title is literally "Support for libgpiod as a pin factory?".

>  Testing is done with a [mock factory](https://github.com/gpiozero/gpiozero/blob/master/gpiozero/pins/mock.py) for the device classes, but with [physical tests](https://github.com/gpiozero/gpiozero/blob/6bfab8aa52c4098503589c3cd0bff7371a598ff4/tests/test_real_pins.py#L32) for the pin factories themselves (because there's no other reliable way of testing the libraries that just bash the GPIO registers). It's one of those things I've been promising myself to set up a runner for, but never find the time to manage.

I've pushed where I'm at into a [branch](https://github.com/warthog618/gpiozero/tree/gpiod).  I haven't run it through your test suite yet, and I don't even recall how thoroughly I tested it since the last round of changes before I committed it, so I'm not about to make a PR yet, but it is there in case anyone is interested or I get hit by a bus.

I do recall that I found a kernel bug while testing, so that was fun.  Specifically if you enable debounce without edge detection, then subsequently enable edge detection, then the kernel event buffer is not correctly initialised and the events emitted contain garbage.  You still get an event, just with no useful information in it.  Yay.    The fix for that made it into 6.9 (just barely) and will be backported to older/stable kernels.  Not sure how long that will take to trickle through to Pi kernels.  Anyone know how long it takes for mainline bug fixes to be backported to Pi kernels?

In the meantime, enabling debounce like this:

```
factory = GpiodFactory(chip=0)
button = Button(22, pin_factory=factory, bounce_time=0.01)
```

triggers the bug, as the debounce happens to be applied before the edge detection.
A workaround would be for the pin factory to first disable debounce when enabling edge detection, if it were set, then configure both at once.
Though that becomes pointless noise once the kernel is patched, so I would rather not....
Is asking for any impacted users to patch their kernel too much to ask?

**Comment by Gadgetoid** [comment]
> `gpiod.line.Value` doesn't cast to `bool`

This really grinds my gears! Is it something that can be coerced into an upstream fix without breakages do you think?

**Comment by warthog618** [comment]
> > `gpiod.line.Value` doesn't cast to `bool`
> 
> This really grinds my gears! Is it something that can be coerced into an upstream fix without breakages do you think?

I was planning on sending a patch to Bart to see what he thinks.  I don't see it breaking anything - as it stands it always casts to True, so no one can be using it in practice.  I definitely see it as a bug fix.

**Comment by warthog618** [comment]
Last I checked, my **gpiod** pin factory branch is passing all tests, and the kernel bug I found has now been patched in the stable kernel, so I am now considering a pull request.  But, not being familiar with **gpiozero**, I would appreciate some guidance as to what else you would expect in a pull request other than the pin factory itself, e.g. updating the _gpiozero_pin_factories_, updating documentation etc.

Any pointers appreciated.

**Comment by Jeducious** [comment]
Hey :) just wanted to flag and ask for help. Installing gpiozero on rp5 (Trixie OS) and installing into a venv causes eventually leads to a python error when trying to set the pin factory, my script fails with:

```Python
.venv/lib/python3.13/site-packages/gpiozero/devices.py:270: in ensure_pin_factory
    Device.pin_factory = Device._default_pin_factory()
```
I followed the rabbit down the hole and ended up here as I noticed the system python runs just fine. The difference is the system python has lgpio. I tried to install it into the venv with pip, but got this error:

```bash
Building wheels for collected packages: lgpio
  Building wheel for lgpio (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for lgpio (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      running bdist_wheel
      running build
      running build_py
      running build_ext
      building '_lgpio' extension
      swigging lgpio.i to lgpio_wrap.c
      swig -python -o lgpio_wrap.c lgpio.i
      error: command 'swig' failed: No such file or directory
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for lgpio
Failed to build lgpio
ERROR: Failed to build installable wheels for some pyproject.toml based projects (lgpio)
```

I also note that the lgpio's [pypi page](https://pypi.org/project/lgpio/) has this warning:
```
Installation
pypy can not handle Linux libraries properly at the moment.

You need to install the lgpio module from the lg archive.

See http://abyz.me.uk/lg/download.html
```
Following that link presents me with the following instructions;

```
Download & Install
Prerequisites
A few packages are needed during installation of the Python modules.

the SWIG code generator (to build lgpio.py from the C library)
the Python development files (to build lgpio.py)
the Python set up tools (to install lgpio.py and rgpio.py)


These packages may be installed with the following command.

sudo apt install swig python-dev python3-dev
sudo apt install python-setuptools python3-setuptools

Download&Install
The lg archive may be downloaded and installed with the following commands.

wget http://abyz.me.uk/lg/lg.zip
unzip lg.zip
cd lg
make
sudo make install
```

I feel like I am missing something after reading this issue though, is there a way to install gpiozero into a venv that doesn't require building lgpio? I thought perhaps I could create a symlink from the system lgpio to my venv, would that be a potential short term workaround? 

My application rquires PWM as I am controlling three phase DC motors. Are there other pin factories that can install via pip rather than buildling lgpio from source.

Appreciate your time, I'm a newbie to this, a PhD student trying to bulid my thesis experimental rig. I've fallen out of the noob tree and hit every branch on the way down it feels like! XD

**Comment by Gadgetoid** [comment]
Wildcard approach is to create the venv with the --system-site-packages flag so it passes through the existing system packages. Can be as much a blessing as a curse but should work here.

**Comment by Jeducious** [comment]
@Gadgetoid you should get a reward for the fastest round trip time for comments on GH mate ;)
> Wildcard approach is to create the venv with the --system-site-packages flag so it passes through the existing system packages. Can be as much a blessing as a curse but should work here.

Ok, sounds reasonable, only, forgive my ignorance here, doesn't that negate most of the reasons one would want a venv? If all the system packages are passed through, seems like just using the system python env is easier, esp on a pi that only as one purpose?

Or did I misunderstand how the `--system-site-packages` flag works? Like, can I still install new packages in the venv and keep the system python env clean? Or will the setting that flag cause packages installed to the venv with pip to also go into the system lib? Not used that flag before! 

Thanks!

**Comment by Gadgetoid** [comment]
Packages installed in the venv stay in the venv but it does mitigate some of the benefits since your venv is no-longer the true runtime for your project.

lgpio ought to install in the venv if SWIG is available - it’ll be on apt as “swig3” or “swig4” (I’ve lost count) and is a tool to auto-generate Python bindings.

**Comment by Jeducious** [comment]
Hi @Gadgetoid, 

Thanks for the suggestion to try installing swig.

So I have these packages available to me;

```bash
$ sudo apt list | grep swig

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

python3-swiglpk/stable 5.0.10-2+b3 arm64
python3-swiglpk/stable 5.0.10-2+b2 armhf
swig-doc/stable,stable 4.3.0-1 all
swig-examples/stable,stable 4.3.0-1 all
swig-pocketsphinx/stable,stable 0.8+5prealpha+1-15 all
swig-sphinxbase/stable,stable 0.8+5prealpha+1-21 all
swig/stable 4.3.0-1 arm64
swig/stable 4.3.0-1 armhf
```
I tried installing using a naive approach where I just installed with `sudo apt install swig` as there wasn't a `swig3` or `swig4`. 

That didn't work, I got more or less the same error when trying to install lgpio. So I guess it will be the `--system-site-packages` route for me then :S

I also installed the pypi swig package, but to no avail (I installed that in the .venv by the way).

**Comment by lurch** [comment]
I've not tried doing any of this myself, so can't offer any advice regarding building things with SWIG, but I **can** offer one useful tidbit WRT installing GpioZero into a venv...
Note that if you install GpioZero using `pip`, you'll get [GpioZero 2.0.1](https://pypi.org/project/gpiozero/) which still contains bug #1166 , whereas the "system-installed" apt version of `python3-gpiozero` is 2.0.1-0+rpt1+trixie which contains a fix similar to #1181 already applied. This means that the system-installed version of GpioZero works out of the box, but a pip-installed version of GpioZero _might_ not work if you're using lgpio on a Raspberry Pi 5. (Of course I guess you could manually apply #1181 to your pip-installed version of GpioZero if you wanted to!)

**Comment by Jeducious** [comment]
@lurch 
> I've not tried doing any of this myself, so can't offer any advice regarding building things with SWIG, but I **can** offer one useful tidbit WRT installing GpioZero into a venv... Note that if you install GpioZero using `pip`, you'll get [GpioZero 2.0.1](https://pypi.org/project/gpiozero/) which still contains bug [#1166](https://github.com/gpiozero/gpiozero/issues/1166) , whereas the "system-installed" apt version of `python3-gpiozero` is 2.0.1-0+rpt1+trixie which contains a fix similar to [#1181](https://github.com/gpiozero/gpiozero/pull/1181) already applied. This means that the system-installed version of GpioZero works out of the box, but a pip-installed version of GpioZero _might_ not work if you're using lgpio on a Raspberry Pi 5. (Of course I guess you could manually apply [#1181](https://github.com/gpiozero/gpiozero/pull/1181) to your pip-installed version of GpioZero if you wanted to!)

I'm interested in learning about what it would take to apply the patch, I'm just a bit behind on understanding what this fix would do. Seems like it would allow GpioZero to use the native pin factory? Or am I way off there?

**Comment by lurch** [comment]
> I'm interested in learning about what it would take to apply the patch

It'd just be a case of looking at https://github.com/gpiozero/gpiozero/pull/1181/files and then making those same changes to the `lgpio.py` in your pip-installed version of GpioZero. (I've never needed to do this myself, as I've always just used the system-installed `python3-gpiozero` package.)

> I'm just a bit behind on understanding what this fix would do. Seems like it would allow GpioZero to use the native pin factory? Or am I way off there?

As explained in #1166 , in some Linux kernel versions the gpiochip interface on the Raspberry Pi 5 appears as `/dev/gpiochip4`, and on other kernel versions it appears as `/dev/gpiochip0`. And the problem is that `lgpio.py` (i.e. the `LGPIOFactory` class) in the pip-published version of GpioZero assumes that the Pi 5 is _always_ `/dev/gpiochip4`; therefore any GpioZero code using lgpio might not work on the Raspberry Pi 5, depending on which Linux kernel version you're using. #1181 fixes `LGPIOFactory` to try using `/dev/gpiochip0` if `/dev/gpiochip4` doesn't exist.
